_episodes/l1-05-continuous_integration.md

```diff
@@ -24,32 +24,468 @@ keypoints:
+- Billing varies depending on the operating system used by your jobs. Running on Windows
+  or MacOS is more expensive.
+
+[billing]: https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions
+[github-org]: https://www.imperial.ac.uk/admin-services/ict/self-service/research-support/research-support-systems/github/working-with-githubcom/
 
 ## Ways to use CI
 
-- many many actions exist
-- enforce style and checks
-- run builds
-- run tests
-- interface with external systems e.g. publish
-- use GitHub services e.g. Dependabot
+Now that we've set up and configured GitHub Actions, what can we use it for? The [GitHub
+Marketplace] is a good place to get ideas but the number of available actions can be
+overwhelming.
+
+[GitHub MarketPlace]: https://github.com/marketplace?type=actions
+
+### Enforce Style and Formatting
+
+One of the simplest uses of CI is to enforce common style and formatting standards to
+code. The below workflow runs [Flake8] to check that all Python code in the repository
+conforms to the PEP8 style guide. Having this workflow ensures that all code added to
+the repository has a consistent style and appearance.
+
+[Flake8]: https://flake8.pycqa.org/
+
+```yaml
+on:
+  - push
+jobs:
+  qa:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+      - uses: py-actions/flake8@v2
+```
+
+### Build Software
+
+Depending on your project you may have a compile or build step needed to make the
+software usable. An example is given below of building a project using the CMake
+toolchain. Common compilers (e.g. gcc, g++) and tools (e.g. Make) are pre-installed but
+you may need additional setup actions if you have specific requirements for different
+versions.
+
+The value of this kind of workflow is pretty straight forward. You can check that a
+freshly checked out version of your can be successfully built. You can run similar
+builds across a variety of operating systems and compilers to ensure broad
+compatibility.
+
+```yaml
+on:
+  - push
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/actions-setup-cmake@v1
+        with:
+          cmake-version: "3.3"
+      - run: |
+          mkdir build
+          cd build
+          cmake ..
+          make all
+```
+
+### Run Tests
+
+Writing tests is an important best practice for software development. Even better is
+incorporating tests into your CI so you know they pass in a newly checked-out repository
+on another computer.
+
+The below shows an example of runnitg the tests of a Python project:
+
+```yaml
+on:
+  - push
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+      - run: pip install -r requirements-dev.txt
+      - run: pytest .
```

Comment:
Might it be worth using the multiline form of `run` here? I suppose you might also need to explain it to students in this case.

```suggestion
      - run: |
          pip install -r requirements-dev.txt
          pytest .
```

Reply:
Yeah tricky one. On balance I think probably best to avoid complicating things with additional yaml syntax even if that is how I'd probably write it IRL.